### PR TITLE
yaml: replace yaml.safe_load() with CSafeLoader

### PR DESCRIPTION
**Pull request by chipaca**
that is, yaml.safe_load(...) -> yaml.load(..., Loader=yaml.CSafeLoader)
This fixes [lp:1787373](https://bugs.launchpad.net/snapcraft/+bug/1787373).

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [ ] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

the tests fail, but they're failing for me on master as well (I'm probably missing something).

### CODE REVIEW

File: snapcraft/storeapi/_client.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 # Set urllib3's logger to only emit errors, not warnings. Otherwise even
 # retries are printed, and they're nasty.
-logging.getLogger(requests.packages.urllib3.__package__).setLevel(logging.ERROR)
+logging.getLogger(requests.packages.urllib3.__package__).setLevel(logging.DEBUG)
```
Comment (sparkiegeek) [inline snapcraft/storeapi/_client.py]: stray debugging accidentally left in?

**Comment by sparkiegeek** [comment]
Overall this looks good, but I think it's worth extracting a shared `load_yaml` function so that this is easier to identify and have future naive `yaml.load` additions spotted in review

**Comment by chipaca** [comment]
I've been bothered by this all the way to physio and back: I should've called out @sparkiegeek for pointing this whole thing  out to me. Thanks dude.

File: snapcraft/storeapi/_client.py (hunk 1 above)
Comment (chipaca) [reply snapcraft/storeapi/_client.py]: I have no idea what you're talking about :innocent:

**Comment by chipaca** [comment]
@sparkiegeek agreed, a helper function would be better. I'll wait for a go-ahead from the snapcraft gods though.

**Comment by sergiusens** [comment]
Thanks @chipaca 
You might want to add yaml-dev to the snapcraft.yaml here as well. If it uses ctypes you might want to look at patches/ctypes_init.diff too
WRT the loader mechanism that @sparkiegeek mentions, we have indeed already talked about doing so to also wrap the yaml representation bits (used to setup ordered writes for OrderedDict among others) into a single method call instead of doing it globally. If you have the time, that would be a welcoming treat!

**Comment by sergiusens** [comment]
There are a couple of issues here @chipaca, care to take a look?

**Comment by kyrofa** [comment]
I like this idea, but yeah, we've been wanting to extract this behavior into a module anyway, which is a little more than we'd ask you to do @chipaca. I hope you don't mind, but I've just pushed a commit to this PR.

File: snapcraft/yaml_utils.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    return OrderedDict(loader.construct_pairs(node))
+
+
+def _str_presenter(dumper, data):
+    if len(data.splitlines()) > 1:  # check for multiline string
+        return dumper.represent_scalar("tag:yaml.org,2002:str", data, style="|")
+    return dumper.represent_scalar("tag:yaml.org,2002:str", data)
+
+
+yaml.add_representer(str, _str_presenter)
```
Comment (sergiusens) [inline snapcraft/yaml_utils.py]: can we add these to our own private class like https://stackoverflow.com/questions/5121931/in-python-how-can-you-load-yaml-mappings-as-ordereddicts
Affecting yaml globally has the effect of things maybe working and maybe not when running tests depending on when this module is imported

Comment (kyrofa) [reply snapcraft/yaml_utils.py]: Sure, done.

**Comment by sergiusens** [comment]
these errors might mean you probably want a private load and dump for the states like the GlobalState has. This would allow us to use the C dump logic everywhere else.

File: snapcraft/extractors/_metadata.py [hunk 3]
```diff
@@ -32,7 +33,8 @@ def __init__(
         version: str = "",
         grade: str = "",
         icon: str = "",
-        desktop_file_paths: List[str] = None
+        desktop_file_paths: List[str] = None,
+        _data: Dict[str, Union[str, List[str]]] = None,
```
Comment (sergiusens) [inline snapcraft/extractors/_metadata.py]: remove the, at the end of this line to get rid of the flake8 error :-/

File: snapcraft/extractors/_metadata.py [hunk 4]
```diff
@@ -48,10 +47,7 @@ def __init__(
         :param list desktop_file_paths: Extracted desktop file paths
         """  # noqa
 
-        if _data:
-            self._data = _data
-        else:
-            self._data = {}
+        self._data = {}  # type: Dict[str, Union[str, List[str]]]
```
Comment (sergiusens) [inline snapcraft/extractors/_metadata.py]: this is just me being blind, but I really started to prefer using `dict()` for these cases

**Comment by sergiusens** [comment]
Can we apply whatever strategy we would do when moving to spread instead?

**Comment by sergiusens** [comment]
It's finally happening!!!!!

**Comment by sergiusens** [comment]
@kyrofa keep an eye on the weird architectures for this one.

**Pull request merged**